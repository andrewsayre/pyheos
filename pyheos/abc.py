"""Define abstract base classes for HEOS."""

from abc import ABC
from typing import Any


class RemoveHeosFieldABC(ABC):
    """Define an abstract base class that removes the 'heos' from dataclass's fields list to prevent serialization."""

    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
        """Post initialize the player."""
        # Prevent the heos instance from being serialized
        fields = self.__dataclass_fields__.copy()  # type: ignore[has-type] # pylint: disable=access-member-before-definition
        del fields["heos"]
        self.__dataclass_fields__ = fields
