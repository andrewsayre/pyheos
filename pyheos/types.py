"""Define the types module for HEOS."""

from enum import IntEnum, StrEnum


class AddCriteriaType(IntEnum):
    """Define the add to queue options."""

    PLAY_NOW = 1
    PLAY_NEXT = 2
    ADD_TO_END = 3
    REPLACE_AND_PLAY = 4


class ConnectionState(StrEnum):
    """Define the possible connection states."""

    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    RECONNECTING = "reconnecting"


class LineOutLevelType(IntEnum):
    """Define the line out level types."""

    UNKNOWN = 0
    VARIABLE = 1
    FIXED = 2


class VolumeControlType(IntEnum):
    "Define control types."

    UNKNOWN = 0
    NONE = 1
    IR = 2
    TRIGGER = 3
    NETWORK = 4


class NetworkType(StrEnum):
    """Define the network type."""

    WIRED = "wired"
    WIFI = "wifi"
    UNKNOWN = "unknown"


class ControlType(StrEnum):
    """Define the control types."""

    PLAY = "play"
    PAUSE = "pause"
    STOP = "stop"
    PLAY_NEXT = "play_next"
    PLAY_PREVIOUS = "play_previous"


class MediaType(StrEnum):
    """Define the media types."""

    ALBUM = "album"
    ARTIST = "artist"
    CONTAINER = "container"
    DLNA_SERVER = "dlna_server"
    GENRE = "genre"
    HEOS_SERVER = "heos_server"
    HEOS_SERVICE = "heos_service"
    MUSIC_SERVICE = "music_service"
    PLAYLIST = "playlist"
    SONG = "song"
    STATION = "station"


class PlayState(StrEnum):
    """Define the play states."""

    PLAY = "play"
    PAUSE = "pause"
    STOP = "stop"


class SignalType(StrEnum):
    """Define the signal names."""

    PLAYER_EVENT = "player_event"
    GROUP_EVENT = "group_event"
    CONTROLLER_EVENT = "controller_event"
    HEOS_EVENT = "heos_event"


class SignalHeosEvent(StrEnum):
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    USER_CREDENTIALS_INVALID = "usercredentials_invalid"


class RepeatType(StrEnum):
    """Define the repeat types."""

    ON_ALL = "on_all"
    ON_ONE = "on_one"
    OFF = "off"
