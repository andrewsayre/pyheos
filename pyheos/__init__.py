"""pyheos - a library for interacting with HEOS devices."""
from . import const
from .dispatch import Dispatcher
from .heos import Heos
from .player import HeosNowPlayingMedia, HeosPlayer
from .response import CommandError
from .source import HeosSource, InputSource

__all__ = [
    'const',
    'CommandError',
    'Dispatcher',
    'Heos',
    'HeosPlayer',
    'HeosNowPlayingMedia',
    'HeosSource',
    'InputSource'
]
