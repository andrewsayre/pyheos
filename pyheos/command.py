"""Define the HEOS command module."""
from typing import Sequence, Tuple

from . import const
from .player import HeosNowPlayingMedia


class HeosCommands:
    """Define a class that encapsulates well-known commands."""

    def __init__(self, connection):
        """Init the command wrapper."""
        self._connection = connection

    async def heart_beat(self, *, raise_for_result=False) -> bool:
        """Perform heart beat command."""
        response = await self._connection.command(
            const.COMMAND_HEART_BEAT, raise_for_result,
            raise_for_result=raise_for_result)
        return response.result

    async def register_for_change_events(
            self, enable=True, *, raise_for_result=False) -> bool:
        """Enable or disable change event notifications."""
        params = {
            'enable': "on" if enable else "off"
        }
        response = await self._connection.command(
            const.COMMAND_REGISTER_FOR_CHANGE_EVENTS, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_players(self) -> Sequence[dict]:
        """Get players."""
        response = await self._connection.command(
            const.COMMAND_GET_PLAYERS, raise_for_result=True)
        return response.payload

    async def get_player_state(self, player_id: int) -> str:
        """Get the state of the player."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_PLAY_STATE, params, raise_for_result=True)
        return response.get_message('state')

    async def set_player_state(self, player_id: int, state: str,
                               *, raise_for_result=False) -> bool:
        """Set the state of the player."""
        if state not in const.VALID_PLAY_STATES:
            raise ValueError("Invalid play state: " + state)
        params = {
            'pid': player_id,
            'state': state
        }
        response = await self._connection.command(
            const.COMMAND_SET_PLAY_STATE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_now_playing_state(
            self, player_id: int,
            now_playing_media: HeosNowPlayingMedia):
        """Get the now playing media information."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_NOW_PLAYING_MEDIA, params, raise_for_result=True)
        now_playing_media.from_data(response.payload)

    async def get_volume(self, player_id: int) -> int:
        """Get the volume of the player."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_VOLUME, params, raise_for_result=True)
        return int(response.get_message('level'))

    async def set_volume(self, player_id: int, level: int,
                         *, raise_for_result=False) -> bool:
        """Set the volume of the player."""
        if level < 0 or level > 100:
            raise ValueError("'level' must be in the range 0-100")
        params = {
            'pid': player_id,
            'level': level
        }
        response = await self._connection.command(
            const.COMMAND_SET_VOLUME, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_mute(self, player_id: str) -> bool:
        """Get the mute state of the player."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_MUTE, params, raise_for_result=True)
        return response.get_message('state') == 'on'

    async def set_mute(self, player_id: str, state: bool,
                       *, raise_for_result=False) -> bool:
        """Set the mute state of the player."""
        params = {
            'pid': player_id,
            'state': "on" if state else "off"
        }
        response = await self._connection.command(
            const.COMMAND_SET_MUTE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def volume_up(self, player_id: int,
                        step: int = const.DEFAULT_STEP,
                        *, raise_for_result=False) -> bool:
        """Increase the volume level."""
        if step < 1 or step > 10:
            raise ValueError("'step' must be in the range 1-10")
        params = {
            'pid': player_id,
            'step': step
        }
        response = await self._connection.command(
            const.COMMAND_VOLUME_UP, params,
            raise_for_result=raise_for_result)
        return response.result

    async def volume_down(self, player_id: int,
                          step: int = const.DEFAULT_STEP,
                          *, raise_for_result=False) -> bool:
        """Increase the volume level."""
        if step < 1 or step > 10:
            raise ValueError("'step' must be in the range 1-10")
        params = {
            'pid': player_id,
            'step': step
        }
        response = await self._connection.command(
            const.COMMAND_VOLUME_DOWN, params,
            raise_for_result=raise_for_result)
        return response.result

    async def toggle_mute(self, player_id: int,
                          *, raise_for_result=False) -> bool:
        """Toggle the mute state.."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_TOGGLE_MUTE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_play_mode(self, player_id: int) -> Tuple[str, bool]:
        """Get the current play mode."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_PLAY_MODE, params, raise_for_result=True)
        repeat = response.get_message('repeat')
        shuffle = response.get_message('shuffle') == 'on'
        return repeat, shuffle

    async def set_play_mode(self, player_id: int, repeat: str,
                            shuffle: bool, *, raise_for_result=False):
        """Set the current play mode."""
        if repeat not in const.VALID_REPEAT_MODES:
            raise ValueError("Invalid repeat mode: " + repeat)
        params = {
            'pid': player_id,
            'repeat': repeat,
            'shuffle': 'on' if shuffle else 'off'
        }
        response = await self._connection.command(
            const.COMMAND_SET_PLAY_MODE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def clear_queue(self, player_id: int,
                          *, raise_for_result=False) -> bool:
        """Clear the queue."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_CLEAR_QUEUE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_next(self, player_id: int,
                        *, raise_for_result=False) -> bool:
        """Play next."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_PLAY_NEXT, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_previous(self, player_id: int,
                            *, raise_for_result=False) -> bool:
        """Play next."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_PLAY_PREVIOUS, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_music_sources(self) -> Sequence[dict]:
        """Get available music sources."""
        response = await self._connection.command(
            const.COMMAND_BROWSE_GET_SOURCES, raise_for_result=True)
        return response.payload

    async def browse(self, source_id: int) -> Sequence[dict]:
        """Browse a music source."""
        params = {
            'sid': source_id
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_BROWSE, params, raise_for_result=True)
        return response.payload

    async def play_input(self, player_id: int, input_name: str, *,
                         source_player_id: int = None,
                         raise_for_result=False) -> bool:
        """Play the specified input source."""
        if input_name not in const.VALID_INPUTS:
            raise ValueError("Invalid input name: " + input_name)
        params = {
            'pid': player_id,
            'spid': source_player_id or player_id,
            'input': input_name
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_PLAY_INPUT, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_preset(self, player_id: int, preset: int,
                          *, raise_for_result=False) -> bool:
        """Play the specified preset by 1-based index."""
        if preset < 1:
            raise ValueError("Invalid preset: " + str(preset))
        params = {
            'pid': player_id,
            'preset': preset
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_PLAY_PRESET, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_stream(self, player_id: int, url: str,
                          *, raise_for_result=False) -> bool:
        """Play the specified URL."""
        params = {
            'pid': player_id,
            'url': url
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_PLAY_STREAM, params,
            raise_for_result=raise_for_result)
        return response.result
