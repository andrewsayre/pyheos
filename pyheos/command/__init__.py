"""Define the HEOS command module."""

from typing import Final

# Browse commands
COMMAND_BROWSE_GET_SOURCES: Final = "browse/get_music_sources"
COMMAND_BROWSE_BROWSE: Final = "browse/browse"
COMMAND_BROWSE_PLAY_INPUT: Final = "browse/play_input"
COMMAND_BROWSE_PLAY_PRESET: Final = "browse/play_preset"
COMMAND_BROWSE_PLAY_STREAM: Final = "browse/play_stream"
COMMAND_BROWSE_ADD_TO_QUEUE: Final = "browse/add_to_queue"

# Player commands
COMMAND_GET_PLAYERS: Final = "player/get_players"
COMMAND_GET_PLAY_STATE: Final = "player/get_play_state"
COMMAND_SET_PLAY_STATE: Final = "player/set_play_state"
COMMAND_GET_NOW_PLAYING_MEDIA: Final = "player/get_now_playing_media"
COMMAND_GET_VOLUME: Final = "player/get_volume"
COMMAND_SET_VOLUME: Final = "player/set_volume"
COMMAND_GET_MUTE: Final = "player/get_mute"
COMMAND_SET_MUTE: Final = "player/set_mute"
COMMAND_VOLUME_UP: Final = "player/volume_up"
COMMAND_VOLUME_DOWN: Final = "player/volume_down"
COMMAND_TOGGLE_MUTE: Final = "player/toggle_mute"
COMMAND_GET_PLAY_MODE: Final = "player/get_play_mode"
COMMAND_SET_PLAY_MODE: Final = "player/set_play_mode"
COMMAND_CLEAR_QUEUE: Final = "player/clear_queue"
COMMAND_PLAY_NEXT: Final = "player/play_next"
COMMAND_PLAY_PREVIOUS: Final = "player/play_previous"
COMMAND_PLAY_QUICK_SELECT: Final = "player/play_quickselect"
COMMAND_SET_QUICK_SELECT: Final = "player/set_quickselect"
COMMAND_GET_QUICK_SELECTS: Final = "player/get_quickselects"

# Group commands
COMMAND_GET_GROUPS: Final = "group/get_groups"
COMMAND_SET_GROUP: Final = "group/set_group"
COMMAND_GET_GROUP_VOLUME: Final = "group/get_volume"
COMMAND_SET_GROUP_VOLUME: Final = "group/set_volume"
COMMAND_GET_GROUP_MUTE: Final = "group/get_mute"
COMMAND_SET_GROUP_MUTE: Final = "group/set_mute"
COMMAND_GROUP_TOGGLE_MUTE: Final = "group/toggle_mute"
COMMAND_GROUP_VOLUME_UP: Final = "group/volume_up"
COMMAND_GROUP_VOLUME_DOWN: Final = "group/volume_down"

# System commands
COMMAND_REGISTER_FOR_CHANGE_EVENTS: Final = "system/register_for_change_events"
COMMAND_HEART_BEAT: Final = "system/heart_beat"
COMMAND_ACCOUNT_CHECK: Final = "system/check_account"
COMMAND_REBOOT: Final = "system/reboot"
COMMAND_SIGN_IN: Final = "system/sign_in"
COMMAND_SIGN_OUT: Final = "system/sign_out"
