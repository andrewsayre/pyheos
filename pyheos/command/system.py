"""
Define the system command module.

This module creates HEOS system commands.

Commands not implemented:
    4.1.7 Prettify JSON response
        Helper command to prettify JSON response when user is running CLI controller through telnet.
        This command will not be implemented in the library.
"""

from pyheos import command, const
from pyheos.message import HeosCommand


class SystemCommands:
    """Define functions for creating system commands."""

    @staticmethod
    def register_for_change_events(enable: bool) -> HeosCommand:
        """Register for change events.

        References:
            4.1.1 Register for Change Events"""
        return HeosCommand(
            command.COMMAND_REGISTER_FOR_CHANGE_EVENTS,
            {const.ATTR_ENABLE: const.VALUE_ON if enable else const.VALUE_OFF},
        )

    @staticmethod
    def check_account() -> HeosCommand:
        """Create a check account command.

        References:
            4.1.2 HEOS Account Check"""
        return HeosCommand(command.COMMAND_ACCOUNT_CHECK)

    @staticmethod
    def sign_in(username: str, password: str) -> HeosCommand:
        """Create a sign in command.

        References:
            4.1.3 HEOS Account Sign In"""
        return HeosCommand(
            command.COMMAND_SIGN_IN,
            {const.ATTR_USER_NAME: username, const.ATTR_PASSWORD: password},
        )

    @staticmethod
    def sign_out() -> HeosCommand:
        """Create a sign out command.

        References:
            4.1.4 HEOS Account Sign Out"""
        return HeosCommand(command.COMMAND_SIGN_OUT)

    @staticmethod
    def heart_beat() -> HeosCommand:
        """Create a heart beat command.

        References:
            4.1.5 HEOS System Heart Beat"""
        return HeosCommand(command.COMMAND_HEART_BEAT)

    @staticmethod
    def reboot() -> HeosCommand:
        """Create a reboot command.

        References:
            4.1.6 HEOS Speaker Reboot"""
        return HeosCommand(command.COMMAND_REBOOT)
