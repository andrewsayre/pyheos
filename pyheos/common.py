"""Define the common module."""

from typing import cast

from pyheos.const import TARGET_VERSION


def is_supported_version(
    version: str | None, min_version: str = TARGET_VERSION
) -> bool:
    """Check if a version is supported.

    Args:
        version (str): The version to check.
        min_version (str): The minimum version.

    Returns:
        bool: True if the version is supported, False otherwise.
    """
    if version is None:
        return False
    try:
        sem_ver = get_sem_ver(version)
    except ValueError:
        return False
    min_sem_ver = get_sem_ver(min_version)
    for i, ver in enumerate(sem_ver):
        if ver < min_sem_ver[i]:
            return False
        if ver > min_sem_ver[i]:
            return True
    return True


def get_sem_ver(version: str) -> tuple[int, int, int]:
    """Get the semantic version from a string.

    Args:
        version (str): The version string.

    Returns:
        tuple: The semantic version as a tuple.
    """
    return cast(tuple[int, int, int], tuple(map(int, version.split(".", 3))))
