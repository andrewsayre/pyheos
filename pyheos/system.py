"""Define the System module."""

from dataclasses import dataclass
from functools import cached_property

from pyheos import const


@dataclass(frozen=True)
class HeosHost:
    """Represents a HEOS host.

    This class is used to store information about a HEOS host and is not tracked or updated once created.
    """

    name: str
    model: str
    serial: str | None
    version: str
    ip_address: str
    network: str

    @classmethod
    def from_data(cls, data: dict[str, str]) -> "HeosHost":
        """Create a HeosHost object from a dictionary.

        Args:
            data (dict): The dictionary to create the HeosHost object from.

        Returns:
            HeosHost: The created HeosHost object.
        """
        return HeosHost(
            data[const.ATTR_NAME],
            data[const.ATTR_MODEL],
            data.get(const.ATTR_SERIAL),
            data[const.ATTR_VERSION],
            data[const.ATTR_IP_ADDRESS],
            data[const.ATTR_NETWORK],
        )


@dataclass
class HeosSystem:
    """Represents information about a HEOS system.

    This class is used to store information about a HEOS system and is not tracked or updated once created.
    """

    signed_in_username: str | None
    host: HeosHost
    hosts: list[HeosHost]

    @property
    def is_signed_in(self) -> bool:
        """Return whether the system is signed in."""
        return self.signed_in_username is not None

    @cached_property
    def preferred_hosts(self) -> list[HeosHost]:
        """Return the preferred hosts."""
        return list(
            [host for host in self.hosts if host.network == const.NETWORK_TYPE_WIRED]
        )

    @cached_property
    def connected_to_preferred_host(self) -> bool:
        """Return whether the system is connected to a host."""
        return self.host in self.preferred_hosts
