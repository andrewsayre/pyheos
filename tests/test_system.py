"""Tests for the system module."""

import itertools

from syrupy.assertion import SnapshotAssertion

from pyheos.system import HeosHost, HeosSystem
from pyheos.types import NetworkType


def test_system_preferred_hosts(snapshot: SnapshotAssertion) -> None:
    """Test the preferred hosts property."""
    serials = ["serial", None]
    ip_address = [True, False]
    network = [NetworkType.WIRED, NetworkType.WIFI]
    supported = [True, False]

    hosts: list[HeosHost] = []
    for index, item in enumerate(
        list(itertools.product(serials, ip_address, network, supported))
    ):
        hosts.insert(
            0,
            HeosHost(
                f"name{index}",
                "model",
                item[0],
                "1.0.0",
                f"127.0.0.{index}" if item[1] else None,
                item[2],
                item[3],
            ),
        )

    system = HeosSystem("username", hosts[0], hosts)

    assert system == snapshot


def test_system_preferred_hosts_with_invalid(snapshot: SnapshotAssertion) -> None:
    """Test the preferred hosts property with an invalid ip address."""
    network = [NetworkType.WIRED, NetworkType.WIFI]
    hosts: list[HeosHost] = []
    for index, item in enumerate(network):
        hosts.insert(
            0,
            HeosHost(f"name{index}", "model", None, "1.0.0", "invalid", item, True),
        )
    system = HeosSystem("username", hosts[0], hosts)

    assert system == snapshot
