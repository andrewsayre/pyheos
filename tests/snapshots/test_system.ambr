# serializer version: 1
# name: test_system_preferred_hosts
  dict({
    'connected_to_preferred_host': False,
    'host': dict({
      'ip_address': None,
      'model': 'model',
      'name': 'name15',
      'network': <NetworkType.WIFI: 'wifi'>,
      'preferred_host': False,
      'serial': None,
      'supported_version': False,
      'version': '1.0.0',
    }),
    'hosts': list([
      dict({
        'ip_address': '127.0.0.0',
        'model': 'model',
        'name': 'name0',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': True,
        'serial': 'serial',
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.1',
        'model': 'model',
        'name': 'name1',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.2',
        'model': 'model',
        'name': 'name2',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.3',
        'model': 'model',
        'name': 'name3',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.8',
        'model': 'model',
        'name': 'name8',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': True,
        'serial': None,
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.9',
        'model': 'model',
        'name': 'name9',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.10',
        'model': 'model',
        'name': 'name10',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.11',
        'model': 'model',
        'name': 'name11',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name15',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name14',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name13',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name12',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': False,
        'serial': None,
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name7',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name6',
        'network': <NetworkType.WIFI: 'wifi'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name5',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': False,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': None,
        'model': 'model',
        'name': 'name4',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': False,
        'serial': 'serial',
        'supported_version': True,
        'version': '1.0.0',
      }),
    ]),
    'is_signed_in': True,
    'preferred_hosts': list([
      dict({
        'ip_address': '127.0.0.0',
        'model': 'model',
        'name': 'name0',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': True,
        'serial': 'serial',
        'supported_version': True,
        'version': '1.0.0',
      }),
      dict({
        'ip_address': '127.0.0.8',
        'model': 'model',
        'name': 'name8',
        'network': <NetworkType.WIRED: 'wired'>,
        'preferred_host': True,
        'serial': None,
        'supported_version': True,
        'version': '1.0.0',
      }),
    ]),
    'signed_in_username': 'username',
  })
# ---
